content/2016-03-29-compiler-flags.rst

```diff
@@ -0,0 +1,275 @@
+
+As Pythran users generally don't want the debug info on the generated native
+code, we chose to strip them by default, using the linker flag
+``-Wl,-strip-all`` that removes all symbol informations, including debug
+symbols.
+
+A Step further: Default Symbol visibility
+=========================================
+
+While we're at it, let's call ``nm`` to check if any symbol remains in the
+binary. After all, the Python interpreter still needs some of them to load the
+native extension!
+
+.. code:: sh
+
+    > nm -C -D cda.so
+    [...] skipping > 900 entries
+    000000000001ed00 u nt2::ext::implement<nt2::tag::rem_pio2_ (boost::dispatch::meta::scalar_<boost::dispatch::meta::double_<double> >, boost::dispatch::meta::scalar_<boost::dispatch::meta::double_<double> >, boost::dispatch::meta::scalar_<boost::dispatch::meta::double_<double> >), boost::dispatch::tag::cpu_, void>::__kernel_rem_pio2(double*, double*, int, int, int, int const*)::PIo2
+    000000000001edc0 u nt2::ext::implement<nt2::tag::rem_pio2_ (boost::dispatch::meta::scalar_<boost::dispatch::meta::double_<double> >, boost::dispatch::meta::scalar_<boost::dispatch::meta::double_<double> >, boost::dispatch::meta::scalar_<boost::dispatch::meta::double_<double> >), boost::dispatch::tag::cpu_, void>::__ieee754_rem_pio2(double, double*)::two_over_pi
+    000000000001ed40 u nt2::ext::implement<nt2::tag::rem_pio2_ (boost::dispatch::meta::scalar_<boost::dispatch::meta::double_<double> >, boost::dispatch::meta::scalar_<boost::dispatch::meta::double_<double> >, boost::dispatch::meta::scalar_<boost::dispatch::meta::double_<double> >), boost::dispatch::tag::cpu_, void>::__ieee754_rem_pio2(double, double*)::npio2_hw
+
+I can tell you Python is *not* using nt2 dispatch mechanism to load native
+extensions. Again, the default compiler settings are responsible for this
+noise, and the relevant compiler flag is ``-fvisibility=hidden`` that tells the
+compiler than only the functions flagged with a special attribute are part of
+the external ABI, the other ones are not exported. As Python uses a single
+entry point to load Pythran modules, namely ``PyInit_cda`` for Python3 modules
+and ``initcda`` for Python2 modules [1]_, one can add the ``__attribute__
+((visibility("default")))`` on this symbol and it will be the only exported
+one. This slightly impacts the code size and eventually gives the compiler more
+optimization opportunities, but nothing significant there (131K), apart the pleasure
+of generating cleaner binaries. That's also going to be the default for next
+Pythran version.
+
+Out of chance: getting faster binaries
+======================================
+
+In the (huge) info pages of GCC, near the doc of ``-fvisibility=hidden``,
+there's this (GCC only) compiler flag, ``-fwhole-program`` that implements some
+kind of Link Time Optimization, in the sense that it tells the compiler to
+consider the current compilation unit (or code) as a whole program. ``main`` is
```

Comment:
Suggestion en remplacement de "``main`` is ..." : As specified in the GCC man page, "All public functions and variables with the exception of "main" and those merged by attribute "externally_visible" become static functions and in effect are optimized more aggressively by interprocedural optimizers.", which basically means that every function is considered static except for "main" and the ones that are explicitly told not to be. This allows the compiler for instance to remove functions that are always inlined, and thus win space.